src/Lean/Elab/PreDefinition/Structural/Main.lean

```diff
@@ -194,6 +194,9 @@ def structuralRecursion (preDefs : Array PreDefinition) (termMeasure?s : Array (
     addSmartUnfoldingDef preDef recArgPos
     markAsRecursive preDef.declName
     generateEagerEqns preDef.declName
+  for preDef in preDefs do
+    -- must happen in separate loop so realizations can see eqnInfos of all other preDefs
+    enableRealizationsForConst preDef.declName
```

Comment:
Above (`Mutual.lean`) you do this before the eager generation of eqns. Does that matter?

Reply:
Good catch, this will make `generateEagerEqns` fail once we use `realizeConst` for structural rec. Though I'm getting more convinced we can simply remove `generateEagerEqns` once that move is finished.

Reply:
That would be even better! Although it’s not quite true: while your env-rewinding does the right thing wrt. to `Options` when the definition ends up being realized in the same module, it won’t when the realization happens in a different module.

Maybe we can just get rid of the `backwards` options again; I doubt anyone is using them.

Reply:
I believe using any such option in a module other than the one defining the decl should be prohibited anyway: it could lead to unimportable module diamonds because of conflicting realizations.

Reply:
The environment you use for imported realizeable consts is the pristine environment from right after the `import`, but without any options, right?

There is still the issue with
```
-- A.lean
set_option backward.eqns.deepRecursiveSplit false in
def foo := …

-- no realization of foo.eq_1 in A.lean
```
and
```
-- B.lean
import A

… rw [foo] …
```
Without `generateEagerEqns` the `foo.eq` would be generated in `B.lean` with default options, and the `set_option backward.eqns.deepRecursiveSplit false` would be ineffectual.